### PR TITLE
User-Agent CLI/version header

### PR DESCRIPTION
**Pull request by drcariel**
#2568

### CODE REVIEW

**Comment by drcariel** [comment]
Looks like this in action:
```
"User-Agent": [
    "OpenWhisk-CLI/2017-08-09T19:14:44+00:00"
  ]
```

File: tools/cli/go-whisk-cli/main.go [hunk 1]
```diff
@@ -48,6 +48,7 @@ func init() {
 
     // Rest of CLI uses the Properties struct, so set the build time there
     commands.Properties.CLIVersion = CLI_BUILD_TIME
+    os.Setenv("CLI_Version", CLI_BUILD_TIME)
```
Comment (mdeuser) [inline tools/cli/go-whisk-cli/main.go]: avoid using "global" style variables.  instead set a value on the http config object during it's initialization

Comment (drcariel) [inline tools/cli/go-whisk-cli/main.go]: I was attempting this at first, but isn't as simple. The CLIVersion is intentionally omitted from that config. I assume for good reason?

Comment (drcariel) [inline tools/cli/go-whisk-cli/main.go]: I would either need to inject the CLIVersion properly into ```cli/go-whisk-cli/commands/property.go``` then import this into ```/cli/go-whisk/whisk/client.go```, or just import ```cli/go-whisk-cli/main.go``` into ```/cli/go-whisk/whisk/client.go```.

Not the greatest options imo

Comment (mdeuser) [inline tools/cli/go-whisk-cli/main.go]: maybe this will work.
- add AgentHeader property to the whisk.Config struct
- in commands.go (line 54 or so), set the AgentHeader value in the client configuration init

**Comment by drcariel** [comment]
Thanks for the input @mdeuser , not sure why I found it so difficult the first time round.

**Comment by drcariel** [comment]
PG2~1935

File: tools/cli/go-whisk/whisk/api.go [hunk 2]
```diff
@@ -43,7 +43,6 @@ type ApiListRequestOptions struct {
     Skip            int       `url:"skip"`
     Docs            bool      `url:"docs,omitempty"`
 }
-//type ApiListResponse RetApiArray
 type ApiListResponse RetApiArray
```
Comment (drcariel) [inline tools/cli/go-whisk/whisk/api.go]: Some lingering cleanup from the experimental-api removal PR

**Comment by drcariel** [comment]
PG2~1937

File: core/controller/src/main/scala/whisk/core/controller/WebActions.scala [hunk 3]
```diff
@@ -326,7 +326,7 @@ trait WhiskWebActionsApi
     private val defaultCorsResponse = List(
         `Access-Control-Allow-Origin`(AllOrigins),
         `Access-Control-Allow-Methods`(OPTIONS, GET, DELETE, POST, PUT, HEAD, PATCH),
-        `Access-Control-Allow-Headers`(`Authorization`.name, `Content-Type`.name))
+        `Access-Control-Allow-Headers`(`Authorization`.name, `Content-Type`.name, `User-Agent`.name))
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/WebActions.scala]: do we want this for a web action?

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala [hunk 4]
```diff
@@ -121,7 +121,7 @@ protected[controller] object RestApiCommons {
  */
 protected[controller] trait RespondWithHeaders extends Directives {
     val allowOrigin = `Access-Control-Allow-Origin`(AllOrigins)
-    val allowHeaders = `Access-Control-Allow-Headers`("Authorization", "Content-Type")
+    val allowHeaders = `Access-Control-Allow-Headers`("Authorization", "Content-Type", "User-Agent")
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: what is the value of the controller accepting the user-agent?
should nginx just drop this header?

File: core/controller/src/main/scala/whisk/core/controller/WebActions.scala (hunk 3 above)
Comment (mdeuser) [inline core/controller/src/main/scala/whisk/core/controller/WebActions.scala]: imho, i think giving web actions visibility into the agent header is a good thing

Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/WebActions.scala]: But the controller does nothing with these user agents for POST. And accepting it here for web actions is just for preflight - I suppose it's fine (but for the cli this has no value there). If you want to keep it I think we need a doc update.

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala (hunk 4 above)
Comment (mdeuser) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: if the controller log the user agent header, log queries can see which clients and which client versions are accessing the controller.

Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: the controller does not log user agents - i don't think it should/needs to.
you get that from nginx.

File: core/controller/src/main/scala/whisk/core/controller/WebActions.scala (hunk 3 above)
Comment (drcariel) [inline core/controller/src/main/scala/whisk/core/controller/WebActions.scala]: rest_api.md has been updated

**Comment by drcariel** [comment]
PG1~1942

File: tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala [hunk 5]
```diff
@@ -47,94 +47,16 @@ import whisk.core.entity.Subject
...
+    with WskTestHelpers with RestUtil with BeforeAndAfterAll {
 
     val MAX_URL_LENGTH = 8192 // 8K matching nginx default
 
     val wsk = new Wsk
     private implicit val wskprops = WskProps()
     val namespace = wsk.namespace.whois()
 
-    protected val testRoutePath: String
+    protected val testRoutePath: String = "/api/v1/web"
```
Comment (drcariel) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: Is this appropriate here? or will testRoutePath be populated at this point?

**Comment by drcariel** [comment]
PG2~1955

File: tools/cli/go-whisk/whisk/client.go [hunk 6]
```diff
@@ -204,6 +209,8 @@ func (c *Client) NewRequest(method, urlStr string, body interface{}, includeName
         return nil, werr
     }
 
+    req.Header.Add("User-Agent", "OpenWhisk-CLI/" + c.Config.CLIVersion)
```
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/client.go]: Use a constant for the "OpenWhisk-CLI/" string

File: tools/cli/go-whisk/whisk/client.go [hunk 7]
```diff
@@ -126,6 +127,10 @@ func NewClient(httpClient *http.Client, config *Config) (*Client, error) {
         config.Version = "v1"
     }
 
+    if len(config.CLIVersion) == 0 {
+        config.CLIVersion = "not_set"
+    }
+
```
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/client.go]: In your testing did you find this default setting to be needed?  i was thinking this check/set is not necessary due to the way the CLI version string is initialized

**Comment by rabbah** [comment]
I'd like to raise again the question of what we are gaining by changing the preflight responses for webactions or why thosnchsnge has to affect the controller at all.

File: tools/cli/go-whisk/whisk/client.go (hunk 7 above)
Comment (drcariel) [inline tools/cli/go-whisk/whisk/client.go]: didn't check, just followed suit with the others, will test it out

**Comment by mdeuser** [comment]
@rabbah - to clarify your CORS related comment above.  this is in reference to a couple lines that look like
```val allowHeaders = `Access-Control-Allow-Headers`("Authorization", "Content-Type", "User-Agent")```

if this is not place, would web apps issuing cross origin api requests (implemented by web actions) still be allowed to provide a user agent header in the request?

File: tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala [hunk 8]
```diff
@@ -47,94 +47,16 @@ import whisk.core.entity.Subject
  * Tests web actions.
  */
 @RunWith(classOf[JUnitRunner])
-class WskWebActionsTestsV1 extends WskWebActionsTests {
```
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: There are still two versions of web actions. Why remove the tests for both versions?

File: tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala [hunk 9]
```diff
@@ -250,7 +172,7 @@ trait WskWebActionsTests
                 response.statusCode shouldBe 200
                 response.header("Access-Control-Allow-Origin") shouldBe "*"
                 response.header("Access-Control-Allow-Methods") shouldBe "OPTIONS, GET, DELETE, POST, PUT, HEAD, PATCH"
-                response.header("Access-Control-Allow-Headers") shouldBe "Authorization, Content-Type"
+                response.header("Access-Control-Allow-Headers") shouldBe "Authorization, Content-Type, User-Agent"
```
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: Not sure how this is related to the CLI since web actions cannot be invoked using the CLI.

Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: @mdeuser, see above.

Comment (mdeuser) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: true.  not related to cli, per-se, but web apps making api requests having web app specific user agent headers should be supported.  see related question above https://github.com/apache/incubator-openwhisk/pull/2591#issuecomment-322465029

Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: @mdeuser, if the user wants to have access to the user-agent, he or she can already ad a custom CORS header.

Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: @mdeuser, @drcariel see: https://github.com/apache/incubator-openwhisk/blob/master/docs/webactions.md#options-requests.

Comment (drcariel) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: Maybe I am confused, but I thought this user-agent header was for our own tracking data, not any thing that is exposed to the user for customization. Would enforcing this for web actions potentially hinder users from implementing their own user-agent header? and/or also force them to add a user-agent to all their web action calls?

Comment (mdeuser) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: @dubeejw - understood.  just wanted to be clear about the default behavior.  the docs currently do not tell the user that the Access-Control-Allow-Headers preflight response header is added
https://github.com/apache/incubator-openwhisk/blob/master/docs/webactions.md#options-requests

Comment (mdeuser) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: @drcariel - requests to the controller can come from the cli, gui, apps using the rest interface, and web action invocations.  each can and will have a different user agent header, so the controller must be application agnostic wrt user agent header handling.  the controller does not enforce that a user agent header be part of an incoming request.  if the user agent header is included, then the controller will handle it like any other benign header the controller does not explicitly act upon.

Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: @drcariel, can you split the controller changes out into a separate PR since they are unrelated to the CLI changes?

File: docs/rest_api.md [hunk 10]
```diff
@@ -80,11 +80,14 @@ wsk namespace list -v
...
 ```
 REQUEST:
-[GET]	https://openwhisk.ng.bluemix.net/api/v1/namespaces
+[GET]  https://openwhisk.ng.bluemix.net/api/v1/namespaces
 Req Headers
 {
   "Authorization": [
     "Basic XXXYYYY"
+  ],
+  "User-Agent": [
```
Comment (rabbah) [inline docs/rest_api.md]: back out changes to this file?

Comment (drcariel) [inline docs/rest_api.md]: I backed out the changes referencing the rest api requiring User-Agent, but this is an appropriate doc update for the CLI changes in my mind.

Comment (mdeuser) [inline docs/rest_api.md]: this is an example of a cli command with the `-v` which displays request header.  when the cli inserts its own agent-header, this will now appear in the `-v` output.. so, this update looks good.  was there something else you had in mind @rabbah ?

**Comment by mdeuser** [comment]
Fixes #2568 
Fixes #2571

File: tools/cli/go-whisk/whisk/client.go [hunk 11]
```diff
@@ -47,6 +47,7 @@ const (
     DoNotProcessTimeOut = false
     ExitWithErrorOnTimeout = true
     ExitWithSuccessOnTimeout = false
+    UserAgent = "OpenWhisk-CLI/"
```
Comment (mdeuser) [inline tools/cli/go-whisk/whisk/client.go]: @drcariel - after talking with @csantanapr, we think this would be better defined as a client.go externalized variable that could be overridden by client.go consumers.

Comment (dubeejw) [inline tools/cli/go-whisk/whisk/client.go]: Agreed. `go-whisk` does not know that it is part of a CLI.

File: tools/cli/go-whisk/whisk/client.go [hunk 12]
```diff
@@ -47,6 +47,7 @@ const (
     DoNotProcessTimeOut = false
     ExitWithErrorOnTimeout = true
     ExitWithSuccessOnTimeout = false
+    UserAgent = "CloudFunctions-Client"
```
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/client.go]: Use OpenWhisk-Client here or something simliar

File: tools/cli/go-whisk/whisk/client.go [hunk 13]
```diff
@@ -76,6 +77,7 @@ type Config struct {
     Verbose   	bool
     Debug       bool     // For detailed tracing
     Insecure    bool
+    CLIVersion  string
```
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/client.go]: The client should not know about the existence of a CLI as it is an abstraction.

Comment (dubeejw) [reply tools/cli/go-whisk/whisk/client.go]: You can just have a UserAgent value here that can be modified by the CLI.

File: tools/cli/go-whisk-cli/commands/commands.go [hunk 14]
```diff
@@ -60,6 +60,8 @@ func setupClientConfig(cmd *cobra.Command, args []string) (error){
         Version:    Properties.APIVersion,
         Insecure:   flags.global.insecure,
         Host:       Properties.APIHost,
+        CLIVersion: Properties.CLIVersion,
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/commands.go]: How does CLIVersion get set?

File: tools/cli/go-whisk/whisk/client.go [hunk 15]
```diff
@@ -204,6 +210,8 @@ func (c *Client) NewRequest(method, urlStr string, body interface{}, includeName
         return nil, werr
     }
 
+    req.Header.Add("User-Agent", c.Config.UserAgent + "/" + c.Config.CLIVersion)
```
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/client.go]: Probably better to move CLIVersion logic into go-whisk-cli and just set the agent to `c.Config.UserAgent`.

File: tools/cli/go-whisk/whisk/client.go [hunk 16]
```diff
@@ -126,6 +128,10 @@ func NewClient(httpClient *http.Client, config *Config) (*Client, error) {
         config.Version = "v1"
     }
 
+    if len(config.UserAgent) == 0 {
+        config.UserAgent = "Unknown"
```
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/client.go]: Perhaps we could default to `OpenWhisk-Go-Client` or something like that.

**Comment by mdeuser** [comment]
@drcariel New changes look good to me.  Will also need a peer PR for custom client user-agent value.

File: tools/cli/go-whisk-cli/commands/commands.go [hunk 17]
```diff
@@ -60,6 +60,7 @@ func setupClientConfig(cmd *cobra.Command, args []string) (error){
         Version:    Properties.APIVersion,
         Insecure:   flags.global.insecure,
         Host:       Properties.APIHost,
+        UserAgent:  "OpenWhisk-CLI/" + Properties.CLIVersion,
```
Comment (mdeuser) [inline tools/cli/go-whisk-cli/commands/commands.go]: @drcariel - define the value used by the cli in an exported string variable, then reference it here

File: tools/cli/go-whisk-cli/commands/commands.go [hunk 18]
```diff
@@ -30,6 +30,7 @@ import (
 
 var client *whisk.Client
 const DefaultOpenWhiskApiPath string = "/api"
+var UserAgent string = "OpenWhisk-CLI"
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/commands.go]: Looks like this should be a const.

File: tools/cli/go-whisk/whisk/client.go [hunk 19]
```diff
@@ -126,6 +127,10 @@ func NewClient(httpClient *http.Client, config *Config) (*Client, error) {
         config.Version = "v1"
     }
 
+    if len(config.UserAgent) == 0 || config.UserAgent == "/not set" {
```
Comment (dubeejw) [inline tools/cli/go-whisk/whisk/client.go]: Where does `/not set` come from?

Comment (drcariel) [reply tools/cli/go-whisk/whisk/client.go]: the CLI version, In my testing I saw the plugin will show `OW-Plugin/not set`, so I added it here to cover the missing UserAgent case, aka "/not set"

File: tools/cli/go-whisk-cli/commands/commands.go (hunk 18 above)
Comment (drcariel) [reply tools/cli/go-whisk-cli/commands/commands.go]: not a const so that it can be overwritten by the plugin etc

File: tools/cli/go-whisk/whisk/client.go (hunk 19 above)
Comment (dubeejw) [reply tools/cli/go-whisk/whisk/client.go]: Do you think that is because the version isn't being set in that scenario? @mdeuser, see the above.

Comment (drcariel) [reply tools/cli/go-whisk/whisk/client.go]: the plugin doesn't use the same gradle build I believe (or maybe I didn't build it correctly), so yes, it is not being set. I was thinking I can check for this prior to appending the cli version in commands.go if we wanna move it.

Comment (mdeuser) [reply tools/cli/go-whisk/whisk/client.go]: from main.go
```
// CLI_BUILD_TIME holds the time of the CLI build.  During gradle builds,
// this value will be overwritten via the command:
//     go build -ldflags "-X main.CLI_BUILD_TIME=nnnnn"   // nnnnn is the new timestamp
var CLI_BUILD_TIME string = "not set"
```

Comment (dubeejw) [reply tools/cli/go-whisk/whisk/client.go]: Might be easier to just set the version in the plugin build.

Comment (mdeuser) [reply tools/cli/go-whisk/whisk/client.go]: yep.. it already is.  the "not set" CLI version is a result of a local build i.e. `go build main.go` without providing the CLI_BUILD_TIME linker flag

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala [hunk 20]
```diff
@@ -314,6 +314,11 @@ class WskBasicUsageTests
             }, waitBeforeRetry = Some(1.second), N = 5)
     }
 
+    it should "include CLI user agent headers with outbound requests" in {
+        val stdout = wsk.cli(Seq("action", "list", "-v", "--auth", wskprops.authKey) ++ wskprops.overrides).stdout
```
Comment (mdeuser) [inline tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: i recommend using the wsk.action.list function.  to do that, augment the ListOrGetFromCollection list() function with a `verbose: Boolean = false` parameter and explicitly pass that to the underlying cli() call.

File: tools/cli/go-whisk/whisk/client.go (hunk 19 above)
Comment (mdeuser) [reply tools/cli/go-whisk/whisk/client.go]: oops. sorry,  as soon as i saw the "not set" i was thinking this was in reference to the cli version...

Comment (mdeuser) [reply tools/cli/go-whisk/whisk/client.go]: sooo.. i don't believe the user-agent var will every have a "not set" value.

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala [hunk 21]
```diff
@@ -314,6 +314,11 @@ class WskBasicUsageTests
...
+        stdout should include regex ("""\bUser-Agent\b": \[(\r\n|\r|\n)\s+"OpenWhisk\-CLI/201.*""".r)
```
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: Little simpler to just check for whitespace instead of the carriage returns and new lines. Also seems like the test will fail in 2020.`"""\bUser-Agent\b": \[\s+"OpenWhisk\-CLI/\d+.*""".r`

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala [hunk 22]
```diff
@@ -119,6 +119,10 @@ class WskBasicUsageTests
             wsk.action.get(fullQualifiedName).stdout should include(s"ok: got action ${packageName}/${actionName}")
     }
 
+    it should "include CLI user agent headers with outbound requests" in {
+        wsk.action.list(verboseFlag = Some(true)).stdout should include regex ("""(?i)User-Agent": \[(\r\n|\r|\n)\s+"OpenWhisk\-CLI/201.*""")
```
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: Little simpler to just check for whitespace instead of the carriage returns and new lines. Also seems like the test will fail in 2020. `"""\bUser-Agent\b": \[\s+"OpenWhisk\-CLI/\d+.*""".r`

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala (hunk 21 above)
Comment (drcariel) [reply tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: yes, you're right about 2020, I chose that to follow suit with the rest of the CLI version tests in this file. I will test a cleaner regex out

File: tests/src/test/scala/common/Wsk.scala [hunk 23]
```diff
@@ -142,17 +142,20 @@ trait ListOrGetFromCollection extends FullyQualifiedNames {
...
         nameSort: Option[Boolean] = None,
-        expectedExitCode: Int = SUCCESS_EXIT)(
+        expectedExitCode: Int = SUCCESS_EXIT,
+        verboseFlag: Option[Boolean] = None)(
             implicit wp: WskProps): RunResult = {
         val params = Seq(noun, "list", resolve(namespace), "--auth", wp.authKey) ++
             { limit map { l => Seq("--limit", l.toString) } getOrElse Seq() } ++
-            { nameSort map { n => Seq("--name-sort") } getOrElse Seq() }
+            { nameSort map { n => Seq("--name-sort") } getOrElse Seq() } ++ 
+            { verboseFlag map { v => Seq("-v")} getOrElse Seq()}
```
Comment (mdeuser) [inline tests/src/test/scala/common/Wsk.scala]: use the cli() function provided `verbose` argument.

Comment (dubeejw) [reply tests/src/test/scala/common/Wsk.scala]: https://github.com/apache/incubator-openwhisk/blob/master/tests/src/test/scala/common/Wsk.scala#L1041

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala (hunk 22 above)
Comment (dubeejw) [reply tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: I see there are other tests that will fail in 2020 as well. We can fix those in a separate PR though.

**Comment by dubeejw** [comment]
```
Syntax of a product token

The product identifier is described in the spec as

     product         = token["/' product-version]
     product-version = token
 
This is fairly self explanatory except for the fact that we have no idea what the syntax of a token is! If you dig around you will find that token is defined in httpbis part1 as,

        token     = 1*tchar 

Having fun yet?  So a token is one or more tchars.  A tchar is defined as,

     tchar     = "!" / "#" / "$" / "%" / "&" / "'" / "*"
                    / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
                    / DIGIT / ALPHA
                    ; any VCHAR, except delimiters

The terms DIGIT, ALPHA and VCHAR are defined in yet another IETF spec here, but what are these delimiters that we are not allowed to use?  The delimiters that are not allowed to appear in a token are defined as:

 (DQUOTE and "(),/:;<=>?@[\]{}")
For those of you still paying attention, you may be saying, but I see semi-colons in user agent strings all the time.  That is because they are in the comment not in the product token.  The “comment” has a different set of syntax rules.  The following is valid example of user-agent that is full of special characters,

user-agent: foo&bar-product!/1.0a$*+ (a;comment,full=of/delimiters@{fun})

```

http://bizcoder.com/the-much-maligned-user-agent-header

**Comment by dubeejw** [comment]
Maybe replace `:` with `-`, or put the timestamp part after the date in a comment. @mdeuser, what do you think?

**Comment by mdeuser** [comment]
@dubeejw - good catch.  there has been talk of adding semantic versioning to the various components, so perhaps this build timestamp is good to put inside a comment - leaving the version value open for a future semantic version value.

**Comment by drcariel** [comment]
@dubeejw and @mdeuser can you show me a desired example? In my head, I am going to update this PR to show the User Agent like so:
```
"User-Agent": [
    "OpenWhisk-CLI/1.0 (2017-08-09T19:14:44+00:00)"
  ]
```
Is that what ya'll are looking for?

**Comment by mdeuser** [comment]
@drcariel - is that error log message now gone?

**Comment by drcariel** [comment]
Where would I see that log? In Kibana?

**Comment by drcariel** [comment]
I could not find any logs pertaining to "Illegal header" in PG3~1051.

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala [hunk 24]
```diff
@@ -119,6 +119,10 @@ class WskBasicUsageTests
             wsk.action.get(fullQualifiedName).stdout should include(s"ok: got action ${packageName}/${actionName}")
     }
 
+    it should "include CLI user agent headers with outbound requests" in {
+        val stdout = wsk.action.list(verbose = true).stdout
+        stdout should include regex ("""\bUser-Agent\b": \[\s+"OpenWhisk\-CLI/1.\d+.*""")
```
Comment (rabbah) [inline tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: since the goal is just to run a cli command with verbose on, i'd rather you don't change the `list` method and install use `wsk.cli(Seq("list"), verbose = true)` directly.

Comment (mdeuser) [reply tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: @rabbah - oh.. i'm wondering why the entity functions don't have a `verbose` pass through to the `wsk.cli` fuction?

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala [hunk 25]
```diff
@@ -120,7 +120,7 @@ class WskBasicUsageTests
     }
 
     it should "include CLI user agent headers with outbound requests" in {
-        val stdout = wsk.cli(Seq("list"), verbose = true).stdout
+        val stdout = wsk.cli(Seq("list", "--auth", wskprops.authKey) ++ wskprops.overrides, verbose = true).stdout
         stdout should include regex ("""\bUser-Agent\b": \[\s+"CloudFunctions\-CLI/1.\d+.*""")
```
Comment (rabbah) [inline tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: CloudFunctions?

Comment (drcariel) [reply tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: Testing something out, will change back shortly.

**Comment by rabbah** [comment]
Is the OS info part of the header?

**Comment by dubeejw** [comment]
@rabbah, @drcariel, @mdeuser, looks like we can easily append the OS to the user-agent if we want to. To do so, import `runtime`, and call `runtime.GOOS`, which will return `windows`, `linux`, or `darwin`.

**Comment by dubeejw** [comment]
The runtime package also provides the ability to retrieve the CPU architecture through `runtime.GOARCH`.

**Comment by mdeuser** [comment]
i'd prefer to do that OS addition in a separate PR.

**Comment by dubeejw** [comment]
I think this one should be ready for merge after Travis checks out.
